src/aarch64/ffi.c

```diff
@@ -728,35 +750,51 @@ ffi_call_int (ffi_cif *cif, void (*fn)(void), void *orig_rvalue,
   else if (flags & AARCH64_RET_NEED_COPY)
     rsize = 16;
 
-  /* Allocate consectutive stack for everything we'll need.
-     The frame uses 40/80 bytes for: lr, fp, rvalue, flags, sp */
-  context = alloca (sizeof(struct call_context) + stack_bytes + sizeof(struct call_frame) + rsize);
-  _Static_assert(sizeof(struct call_context) == CALL_CONTEXT_SIZE, "");
-  _Static_assert(sizeof(struct call_frame) == CALL_FRAME_SIZE, "");
-  stack = context + 1;
-#ifdef __CHERI_PURE_CAPABILITY__
   /*
-   * context is bounded to the alloca size, we want to pass a pointer that
-   * points just after the alloca with the curretn stack bounds.
+   * Compute sub-objects' layouts, i.e. sizes and offsets.
+   *
+   * On CHERI-extended architectures, the allocation is appropriately padded to
+   * use precise bounds for sub-objects, including the stack and on-stack
+   * function arguments referenced by the c9 capability register.
    */
-  stack = __builtin_cheri_address_set(__builtin_cheri_stack_get(), (ptraddr_t)stack);
+  total_length = 0;
+  SUBOBJECT_LAYOUT(total_length, context_length, context_offset,
+    sizeof(*context), struct call_context);
```

Comment:
typeof(*context) then? It's C23, but has existed for a long time as a GNU extension.

Reply:
It seems to work well with llvm-morello. Thanks!